### ISSUE TITLE
file field

### ISSUE THREAD

**Issue by ginad** [jonathangeiger/kohana-jelly#78]
Uploading of file should be done after validation. right now uploading is done inside the set function.
and that's before the item to be uploaded is validated, what if i want to upload image files only?

```
public function set($value)
{
    if ($this->null AND empty($value))
    {
        $value = $this->default;
    }
    return $value;
}
// only upload if it passes the validation
public function save($model, $value, $loaded){
    // upload if value is array
    if(is_array($value)){
        if (upload::valid($value))
                    {
                        if (FALSE !== ($filename = upload::save($value, NULL, $this->path)))
                        {
                                // Chop off the original path
                                $value = str_replace(realpath($this->path).DIRECTORY_SEPARATOR, '', $filename);
                                // Ensure we have no leading slash
                                if (is_string($value))
                                {
                                        $value = trim($filename, DIRECTORY_SEPARATOR);
                                }
                        }else{
                                $value = $this->default;
                        }
                }           
    }

    return $value;
}
```

I also suggest adding restoring the current value if there is no upload done and no default is defined.

**Comment by jonathangeiger** [comment]
This is a good point. I originally put it in set() because I thought it would be useful to have access to the filename immediately, but forgoing the validation process is utterly wrong. 

I'll fix this soon.

**Comment by ginad** [comment]
if you want to have access to the filename then I think we can generate a file name on set then use that file name in the uploads.

ex. 
// on set, generate a file name
public function set($value){
   if ($this->null AND empty($value))
    {
        $value = $this->default;
    }else{
     $this->filename = uniqid() . Inflector::underscore($value['name']);
   }
   return value;
}

// on save upload and save the file

public function save($model,$value,$loaded){
  // upload if value is array
    if(is_array($value)){
        if (upload::valid($value))
                    {
                        if (FALSE !== (upload::save($value, $this->filename, $this->path)))
                        {
                                // Chop off the original path
                                $value = $this->path.DIRECTORY_SEPARATOR.$filename;
                                // Ensure we have no leading slash
                                if (is_string($value))
                                {
                                        $value = trim($filename, DIRECTORY_SEPARATOR);
                                }
                        }else{
                               // $value = $this->default;
                               $value = $this->previews_value;
                        }
                }  
    }

```
return $value;
```

}

**Comment by jonathangeiger** [comment]
Closed by 1941c49c2839a56e823eb5d6b4fa548053528bf1. Thanks ginad!

**Comment by dyip** [comment]
Does this mean we have to save twice for file_fields in order to access the uploaded file information?

For example, we have a table:

create table `images`(
`id` primary key,
`image_name` varchar(127) <----where File_Field is,
`thumbnail_name` varchar(127),
...
);

So if I want to create a thumbnail from the uploaded image, I have to call save() the image object once, and then once I have the filename, create the thumbnail and save the image object again?
